src/python/pants/backend/jvm/tasks/jvm_compile/zinc/zinc_compile.py

```diff
@@ -538,6 +544,38 @@ def compile(
             },
         )()
 
+    def always_do_after_compile(self, ctx):
+        self._pass_diagnostics_to_buildstats(ctx)
+
+    def _aggregate_diagnostics(self, lsp_data):
+        # Note: this is not arbitrary. It is exactly every value that the LSP's DiagnosticSeverity allows.
+        # See https://microsoft.github.io/language-server-protocol/specification#diagnostic
+        counts = {
+            "Error": 0,
+            "Warning": 0,
+            "Information": 0,
+            "Hint": 0,
+        }
+        for published_diagnostics in lsp_data:
+            for diagnostic in published_diagnostics["diagnostics"]:
+                severity = diagnostic["severity"]
+                counts[severity] += 1
+        return counts
+
+    def _pass_diagnostics_to_buildstats(self, ctx):
```

Comment:
`buildstats` is not a name used in the open source repo, use `pants_stats` instead.